cpp/include/raft/sparse/matrix/preprocessing.cuh

```diff
@@ -0,0 +1,568 @@
+    saveFile << fullIdLen << " ";
+    for (int i = 0; i < numFeats; i++) {
+      saveFile << featIdCount[i] << " ";
+    }
+    saveFile.close();
+  }
+}
+
+template <typename ValueType, typename IndexType>
+void SparseEncoder<ValueType, IndexType>::_fit_feats(raft::resources const& handle,
+                                                     IndexType* cols,
+                                                     IndexType* counts,
+                                                     IndexType nnz,
+                                                     IndexType* results)
+{
+  int blockSize  = (nnz < 256) ? nnz : 256;
+  int num_blocks = (nnz + blockSize - 1) / blockSize;
+  raft::sparse::matrix::detail::_scan<<<blockSize, num_blocks>>>(cols, nnz, counts);
+  raft::sparse::matrix::detail::_fit_compute_occurs<<<blockSize, num_blocks>>>(
+    cols, nnz, counts, results, numFeats);
+}
+
+template <typename ValueType, typename IndexType>
+void SparseEncoder<ValueType, IndexType>::_fit(raft::resources const& handle,
+                                               raft::device_vector_view<IndexType, int64_t> rows,
+                                               raft::device_vector_view<IndexType, int64_t> columns,
+                                               raft::device_vector_view<ValueType, int64_t> values,
+                                               int num_rows)
+{
+  numRows += num_rows;
+  IndexType nnz       = values.size();
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+  auto batchIdLen     = raft::make_host_scalar<ValueType>(0);
+  auto values_mat     = raft::make_device_scalar<ValueType>(handle, 0);
+  raft::linalg::mapReduce<ValueType>(values_mat.data_handle(),
+                                     nnz,
+                                     0.0f,
+                                     raft::identity_op(),
+                                     raft::add_op(),
+                                     stream,
+                                     values.data_handle());
+  raft::copy(batchIdLen.data_handle(), values_mat.data_handle(), values_mat.size(), stream);
+  fullIdLen += (int)batchIdLen(0);
+  auto d_rows = raft::make_device_vector<IndexType, int64_t>(handle, nnz);
+  auto d_cols = raft::make_device_vector<IndexType, int64_t>(handle, nnz);
+  auto d_vals = raft::make_device_vector<ValueType, int64_t>(handle, nnz);
+  raft::copy(d_rows.data_handle(), rows.data_handle(), nnz, stream);
+  raft::copy(d_cols.data_handle(), columns.data_handle(), nnz, stream);
+  raft::copy(d_vals.data_handle(), values.data_handle(), nnz, stream);
+  raft::sparse::op::coo_sort(
+    nnz, nnz, nnz, d_cols.data_handle(), d_rows.data_handle(), d_vals.data_handle(), stream);
+  IndexType* counts;
+  cudaMallocManaged(&counts, nnz * sizeof(IndexType));
+  cudaMemset(counts, 0, nnz * sizeof(IndexType));
+  _fit_feats(handle, d_cols.data_handle(), counts, nnz, featIdCount);
+  cudaFree(counts);
+  cudaDeviceSynchronize();
```

Comment:
cudaFree will sync the device implicitly (before releasing a pointer), so there is no need to call `cudaDeviceSynchronize` separately. Furthermore, <only for suggestion>: I believe `cudaMallocManaged` is low performance, maybe you can try pinned host memory via `cudaMallocHost` when you're pretty sure the device memory is not enough.